File: packages/llmix/python/llmix/provider_kwargs.py
"""Provider kwargs injection -- per-provider request mutation before dispatch.

Each provider can define a `transform_kwargs` callback that mutates request
parameters before the API call is made. This keeps provider-specific quirks
(reasoning model parameter stripping, default extra_body injection, etc.)
isolated from the main client code.

Provider-specific behavior is isolated here so the main client dispatch path
can stay provider-neutral.
"""

import re
from typing import Any, Protocol, TypedDict

from llmix.env import get_gpu_base_url

# =============================================================================
# Types
# =============================================================================


class TransformKwargsContext(TypedDict, total=False):
    """Context passed to transform_kwargs callbacks."""

    model: str
    provider: str
    messages: list[dict[str, Any]]
    temperature: float | None
    top_p: float | None
    enable_thinking: bool | None
    provider_options: dict[str, Any]
    base_url: str | None


class TransformKwargsCallback(Protocol):
    """Callable that mutates request kwargs based on provider-specific rules."""

    def __call__(self, ctx: TransformKwargsContext, kwargs: dict[str, Any]) -> dict[str, Any]: ...


# =============================================================================
# Core dispatch
# =============================================================================


def apply_transform_kwargs(ctx: TransformKwargsContext, kwargs: dict[str, Any], callback: TransformKwargsCallback | None) -> dict[str, Any]:
    """Apply a provider's transform_kwargs callback if non-null.

    Returns kwargs unchanged when callback is None.
    """
    if callback is None:
        return kwargs
    return callback(ctx, kwargs)


# =============================================================================
# OpenAI: strip temperature / top_p for reasoning models
# =============================================================================


def _is_reasoning_model(model_id: str) -> bool:
    """Check if model is an OpenAI reasoning model.

    Mirrors logic from model_capabilities._is_reasoning_model without
    pulling in the full types.py import chain.
    # TEMP: regex patch — migrate to config-driven model capabilities (see model-capabilities.json)
    """
    lower = model_id.lower()
    return bool(re.match(r"^o\d", lower)) or lower.startswith(("gpt-5", "codex-", "computer-use"))


def openai_transform_kwargs(ctx: TransformKwargsContext, kwargs: dict[str, Any]) -> dict[str, Any]:
    """Strip temperature and top_p for OpenAI reasoning models.

    Reasoning models (o-series, gpt-5*, codex-, computer-use-)
    require temperature=1 and do not accept top_p.
    """
    model = ctx.get("model", "")
    if not _is_reasoning_model(model):
        return kwargs

    kwargs = dict(kwargs)
    kwargs.pop("temperature", None)
    kwargs.pop("top_p", None)

    # Reasoning models use max_completion_tokens, not max_tokens
    max_tokens = kwargs.pop("max_tokens", None)
    if max_tokens is not None and "max_completion_tokens" not in kwargs:
        kwargs["max_completion_tokens"] = max_tokens

    return kwargs


# =============================================================================
# OpenRouter: inject extra_body.provider.sort = "price"
# =============================================================================

_OPENROUTER_DEFAULT_PROVIDER = {"sort": "price"}


def openrouter_transform_kwargs(ctx: TransformKwargsContext, kwargs: dict[str, Any]) -> dict[str, Any]:
    """Inject default provider sorting config for OpenRouter.

    Configured OpenRouter provider/reasoning options are copied into
    extra_body when the caller did not already provide explicit kwargs.
    Falls back to extra_body.provider.sort = "price".
    """
    kwargs = dict(kwargs)
    provider_options = ctx.get("provider_options") or {}
    openrouter_options = provider_options.get("openrouter") or {}

    extra_body: dict[str, Any] = kwargs.get("extra_body") or {}
    if "provider" not in extra_body:
        configured_provider = openrouter_options.get("provider")
        provider = configured_provider if isinstance(configured_provider, dict) else _OPENROUTER_DEFAULT_PROVIDER
        extra_body = {**extra_body, "provider": dict(provider)}
    if "reasoning" not in extra_body and isinstance(openrouter_options.get("reasoning"), dict):
        extra_body = {**extra_body, "reasoning": dict(openrouter_options["reasoning"])}
    if extra_body:
        kwargs["extra_body"] = extra_body
    return kwargs


# =============================================================================
# Gemini: default thinking_budget=0, override from providerOptions
# =============================================================================


def gemini_transform_kwargs(ctx: TransformKwargsContext, kwargs: dict[str, Any]) -> dict[str, Any]:
    """Set default ThinkingConfig(thinking_budget=0), override from providerOptions.

    Disables thinking by default. If providerOptions.google.thinking_config.
    thinking_budget is set, uses that value instead. The legacy flat
    providerOptions.google.thinking_budget key is also accepted for
    compatibility.
    """
    kwargs = dict(kwargs)

    provider_options = ctx.get("provider_options") or {}
    google_opts = provider_options.get("google") or {}
    raw_thinking_config = google_opts.get("thinking_config") or {}
    thinking_budget = raw_thinking_config.get("thinking_budget")
    if thinking_budget is None:
        thinking_budget = google_opts.get("thinking_budget")

    thinking_config = kwargs.get("thinking_config") or {}
    # Only set default if not already explicitly configured
    if "thinking_budget" not in thinking_config:
        if thinking_budget is None and ctx.get("enable_thinking") is True:
            return kwargs
        if thinking_budget is None:
            thinking_budget = 0
        thinking_config = {**thinking_config, "thinking_budget": thinking_budget}
        kwargs["thinking_config"] = thinking_config

    return kwargs


# =============================================================================
# Sno GPU: construct base URL from providerOptions["sno-gpu"].gpuPath
# =============================================================================


def sno_gpu_transform_kwargs(ctx: TransformKwargsContext, kwargs: dict[str, Any]) -> dict[str, Any]:
    """Construct base URL from providerOptions["sno-gpu"].gpuPath.

    Builds: {base_url}/{gpuPath}/v1 when gpuPath is present.
    Falls back to {base_url}/v1 when gpuPath is absent.

    Also propagates ``enable_thinking`` (from providerOptions["sno-gpu"] or
    common) into ``extra_body`` in BOTH the top-level form (the wrapper
    translates this to ``chat_template_kwargs.enable_thinking``) AND directly
    under ``chat_template_kwargs`` (so direct-vLLM endpoints honor it without
    the wrapper). Without this, the MDA preset's ``enableThinking: false`` is
    silently dropped and the reasoning model burns the whole token budget on
    chain-of-thought, returning empty ``content``.
    """
    kwargs = dict(kwargs)

    provider_options = ctx.get("provider_options") or {}
    sno_gpu_opts = provider_options.get("sno-gpu") or {}
    gpu_path: str | None = sno_gpu_opts.get("gpu_path")

    enable_thinking_opt = sno_gpu_opts.get("enable_thinking")
    common_thinking = ctx.get("enable_thinking")
    if enable_thinking_opt is not None or common_thinking is not None:
        flag = bool(enable_thinking_opt) if enable_thinking_opt is not None else bool(common_thinking)
        extra_body = dict(kwargs.get("extra_body") or {})
        extra_body.setdefault("enable_thinking", flag)
        chat_template_kwargs = dict(extra_body.get("chat_template_kwargs") or {})
        chat_template_kwargs.setdefault("enable_thinking", flag)
        extra_body["chat_template_kwargs"] = chat_template_kwargs
        kwargs["extra_body"] = extra_body

    base_url = ctx.get("base_url") or ""
    base = base_url.rstrip("/")
    if not base:
        base = get_gpu_base_url().rstrip("/")
    # Strip trailing /v1 if present so we can reconstruct cleanly
    if base.endswith("/v1"):
        base = base[:-3]

    if not base.strip():
        raise ValueError("sno-gpu provider requires a non-empty base_url in config or GPU_BASE_URL env var")

    if gpu_path:
        if '..' in gpu_path or not re.match(r'^[a-zA-Z0-9_/-]+$', gpu_path):
            raise ValueError(f"Invalid gpu_path: {gpu_path!r}")
        kwargs["base_url"] = f"{base}/{gpu_path}/v1"
    else:
        kwargs["base_url"] = f"{base}/v1"

    return kwargs


# =============================================================================
# Registry: provider name -> default callback
# =============================================================================

PROVIDER_KWARGS_REGISTRY: dict[str, TransformKwargsCallback] = {
    "openai": openai_transform_kwargs,
    "openrouter": openrouter_transform_kwargs,
    # Legacy DeepSeek provider configs route through OpenRouter.
    "deepseek": openrouter_transform_kwargs,
    "google": gemini_transform_kwargs,
    "gemini": gemini_transform_kwargs,
    "sno-gpu": sno_gpu_transform_kwargs,
}

File: packages/llmix/typescript/src/provider-kwargs.ts
/**
 * Provider kwargs injection -- per-provider request mutation before dispatch.
 *
 * Each provider can define a `transformKwargs` callback that mutates request
 * parameters before the API call is made. This keeps provider-specific quirks
 * (reasoning model parameter stripping, default extra_body injection, etc.)
 * isolated from the main client code.
 *
 * Provider-specific behavior is isolated here so the main client dispatch path
 * can stay provider-neutral.
 */

import { getGpuBaseUrl } from "./env.js";
import { getModelCapabilities } from "./model-capabilities.js";
import type { ProviderOptions } from "./types.js";

const GPU_PATH_PATTERN = /^[a-zA-Z0-9_-]+(\/[a-zA-Z0-9_-]+)*$/;

// =============================================================================
// Types
// =============================================================================

export interface TransformKwargsContext {
  model: string;
  provider: string;
  messages?: unknown[] | undefined;
  temperature?: number | undefined;
  topP?: number | undefined;
  providerOptions?: ProviderOptions | undefined;
  baseUrl?: string | undefined;
  enableThinking?: boolean | undefined;
}

export type TransformKwargsCallback = (
  ctx: TransformKwargsContext,
  kwargs: Record<string, unknown>
) => Record<string, unknown>;

// =============================================================================
// Core dispatch
// =============================================================================

/**
 * Apply a provider's transformKwargs callback if non-null.
 * Returns kwargs unchanged when callback is undefined/null.
 */
export function applyTransformKwargs(
  ctx: TransformKwargsContext,
  kwargs: Record<string, unknown>,
  callback: TransformKwargsCallback | undefined | null
): Record<string, unknown> {
  if (callback == null) {
    return kwargs;
  }
  return callback(ctx, kwargs);
}

// =============================================================================
// OpenAI: strip temperature / top_p for reasoning models
// =============================================================================

/**
 * Strip temperature and top_p for OpenAI reasoning models.
 *
 * Reasoning models (o-series, gpt-5*, codex-, computer-use-)
 * require temperature=1 and do not accept top_p.
 */
export function openaiTransformKwargs(
  ctx: TransformKwargsContext,
  kwargs: Record<string, unknown>
): Record<string, unknown> {
  const capabilities = getModelCapabilities(ctx.model);
  if (!capabilities.isReasoningModel) {
    return kwargs;
  }

  const result = { ...kwargs };
  delete result["temperature"];
  delete result["top_p"];

  // Reasoning models use max_completion_tokens, not max_tokens
  if (result["max_tokens"] !== undefined && result["max_completion_tokens"] === undefined) {
    result["max_completion_tokens"] = result["max_tokens"];
    delete result["max_tokens"];
  }

  return result;
}

// =============================================================================
// OpenRouter: inject extra_body.provider.sort = "price"
// =============================================================================

const OPENROUTER_DEFAULT_PROVIDER = { sort: "price" } as const;

/**
 * Inject default provider sorting config for OpenRouter.
 * Sets extra_body.provider.sort = "price" when not already present.
 */
export function openrouterTransformKwargs(
  ctx: TransformKwargsContext,
  kwargs: Record<string, unknown>
): Record<string, unknown> {
  const result = { ...kwargs };
  const raw = result["extra_body"];
  const extraBody: Record<string, unknown> =
    typeof raw === "object" && raw !== null && !Array.isArray(raw)
      ? (raw as Record<string, unknown>)
      : {};
  const openrouterOptions = ctx.providerOptions?.openrouter;
  if (!("provider" in extraBody)) {
    extraBody["provider"] = openrouterOptions?.provider ?? OPENROUTER_DEFAULT_PROVIDER;
  }
  if (!("reasoning" in extraBody) && openrouterOptions?.reasoning !== undefined) {
    extraBody["reasoning"] = openrouterOptions.reasoning;
  }
  if (Object.keys(extraBody).length > 0) {
    result["extra_body"] = { ...extraBody };
  }
  return result;
}

// =============================================================================
// Gemini: default thinking_budget=0, override from providerOptions
// =============================================================================

/**
 * Set default ThinkingConfig(thinkingBudget=0), override from providerOptions.
 *
 * Disables thinking by default. When ctx.enableThinking is true and no explicit
 * budget is provided, skips injecting a budget so the provider uses its own default.
 * If providerOptions.google.thinkingConfig.thinkingBudget is set, uses that value.
 */
export function geminiTransformKwargs(
  ctx: TransformKwargsContext,
  kwargs: Record<string, unknown>
): Record<string, unknown> {
  const result = { ...kwargs };

  const googleOpts = ctx.providerOptions?.google;
  const explicitBudget = googleOpts?.thinkingConfig?.thinkingBudget;
  // When enableThinking is true and no explicit budget, let the provider decide
  const thinkingBudget: number | undefined =
    explicitBudget ?? (ctx.enableThinking ? undefined : 0);

  const rawTc = result["thinkingConfig"];
  const thinkingConfig: Record<string, unknown> =
    typeof rawTc === "object" && rawTc !== null && !Array.isArray(rawTc)
      ? (rawTc as Record<string, unknown>)
      : {};
  if (!("thinkingBudget" in thinkingConfig) && thinkingBudget !== undefined) {
    result["thinkingConfig"] = { ...thinkingConfig, thinkingBudget };
  }

  return result;
}

// =============================================================================
// Sno GPU: construct base URL and carry thinking settings
// =============================================================================

function asBoolean(value: unknown): boolean | undefined {
  return typeof value === "boolean" ? value : undefined;
}

function asNumber(value: unknown): number | undefined {
  return typeof value === "number" ? value : undefined;
}

/**
 * Construct base URL from providerOptions["sno-gpu"].gpuPath and surface
 * effective thinking settings so dispatchers can forward them consistently.
 *
 * Builds: {base}/{gpuPath}/v1 when gpuPath is present.
 * Falls back to {base}/v1 when gpuPath is absent.
 */
export function snoGpuTransformKwargs(
  ctx: TransformKwargsContext,
  kwargs: Record<string, unknown>
): Record<string, unknown> {
  const result = { ...kwargs };

  const snoGpuOpts = ctx.providerOptions?.["sno-gpu"];
  const gpuPath = snoGpuOpts?.gpuPath;

  let base = (ctx.baseUrl ?? "").replace(/\/+$/, "");
  if (!base) {
    base = (getGpuBaseUrl() ?? "").replace(/\/+$/, "");
  }
  if (base.endsWith("/v1")) {
    base = base.slice(0, -3);
  }

  if (!base) {
    throw new Error(
      "sno-gpu provider requires a non-empty base_url in config or GPU_BASE_URL env var",
    );
  }

  if (gpuPath) {
    // Validate gpuPath: length cap, no traversal, alphanumeric + hyphens/underscores/slashes only
    if (
      gpuPath.length > 256 ||
      gpuPath.includes("..") ||
      !GPU_PATH_PATTERN.test(gpuPath)
    ) {
      throw new Error(
        `Invalid gpu_path: ${JSON.stringify(gpuPath)}. Must be a safe relative path using letters, digits, "_", "-", or "/".`,
      );
    }
    result["baseUrl"] = `${base}/${gpuPath}/v1`;
  } else {
    result["baseUrl"] = `${base}/v1`;
  }

  const enableThinking =
    asBoolean(result["enableThinking"]) ??
    asBoolean(result["enable_thinking"]) ??
    snoGpuOpts?.enableThinking ??
    ctx.enableThinking;
  if (
    enableThinking !== undefined &&
    result["enableThinking"] === undefined &&
    result["enable_thinking"] === undefined
  ) {
    result["enableThinking"] = enableThinking;
  }

  const thinkingBudget =
    asNumber(result["thinkingBudget"]) ??
    asNumber(result["thinking_budget"]) ??
    snoGpuOpts?.thinkingBudget;
  if (
    thinkingBudget !== undefined &&
    result["thinkingBudget"] === undefined &&
    result["thinking_budget"] === undefined
  ) {
    result["thinkingBudget"] = thinkingBudget;
  }

  return result;
}

// =============================================================================
// Registry: provider name -> default callback
// =============================================================================

export const PROVIDER_KWARGS_REGISTRY: Record<string, TransformKwargsCallback> = {
  openai: openaiTransformKwargs,
  openrouter: openrouterTransformKwargs,
  // Legacy DeepSeek configs route through OpenRouter.
  deepseek: openrouterTransformKwargs,
  google: geminiTransformKwargs,
  gemini: geminiTransformKwargs,
  "sno-gpu": snoGpuTransformKwargs,
};
